### ISSUE TITLE
Add ability to create custom relationships between objects of the same type. 

### ISSUE THREAD

**Issue by scetron** [nautobot/nautobot#157]
### Environment
* Python version: 3.x latest
* Nautobot version: 1.0.0b2 (latest) 

### Proposed Functionality

Extend the custom relationship feature to be able to create relationships of the same object type. EG. between dcim.device and dcim.device. 

### Use Case

This would allow to create a simpler type of Virtual Chassis or perhaps a custom relationship for Cisco FEX devices. There may be other creative uses by removing this limitation. 

### Database Changes

Unsure.

### External Dependencies

There should not be any external dependencies.

**Comment by jedelman8** [comment]
This has already come up a few times.  Will definitely require some exploratory efforts here

The use case we've heard already was FEX devices, but would love to hear if there are other use cases, even beyond virtual chassis.

**Comment by bryanward-net** [comment]
The ability to relate Sites to Sites could help with #230 .  Similarly, I'd love to have VLAN Groups be able to inherit from each other so that VLANs that span several "sites" or "buildings" can be made available to all interfaces within a site/building.

**Comment by cmacnevin** [comment]
Agree, nested sites is a long term request of ours which never got traction. But if we can associate prefixes with any object - eg rack groups or regions, then it's not so important to us.

**Comment by ggidofalvy-tc** [comment]
Chiming in on this feature request, this would also open up the possibilities for point-to-multipoint and multipoint-to-multipoint connectivity, if one could define connectivity between interfaces. Probably better suited for a generalized multipoint circuit model, but this could be another approach to modelling the outlined scenario.

**Comment by glennmatthews** [comment]
Based on the best information we have so far, all of the concrete use cases we have examined for same-to-same relationships fall into the "mutually exclusive" category, so for at least the time being, we will *not* be supporting the option for an object to act as both source and destination for the same Relationship.
